File: app/src/main/java/the/mrsmile/foodorderer/activities/OrderActivity.java
package the.mrsmile.foodorderer.activities;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.snackbar.Snackbar;

import java.util.ArrayList;
import java.util.List;

import the.mrsmile.foodorderer.R;
import the.mrsmile.foodorderer.adapters.OrdersAdapter;
import the.mrsmile.foodorderer.models.CategoryItems;
import the.mrsmile.foodorderer.models.OrderItems;

public class OrderActivity extends AppCompatActivity implements OrdersAdapter.onOrderClick {

    private OrdersAdapter adapter;
    private final List<OrderItems> list = new ArrayList<>();
    public static final String Burger = "Burger";
    public static final String Momos = "Momos";
    public static final String PaneerTikka = "Paneer Tikka";
    public static final String burger1 = "Lite Whopper Jr Veg";
    public static final String burger2 = "Crispy Veg Burger";
    public static final String burger3 = "Veg Makhani Burst Burger";
    public static final String burger4 = "Crispy Veg Double Patty Burger";
    public static final String burger5 = "Tikki Twist Burger";
    int burger1id = R.drawable.lite_whooper_jr;
    int burger2id = R.drawable.crispy_veg;
    int burger3id = R.drawable.veg_makhani;
    int burger4id = R.drawable.double_patty;
    int burger5id = R.drawable.tikki_twist;
    String burger1price = "Rs.109";
    String burger2price = "Rs.55";
    String burger3price = "Rs.60";
    String burger4price = "Rs.75";
    String burger5price = "Rs.60";
//    public static final String Pizza = "Pizza";
//    public static final String Chaamp = "Chaamp";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_order);

        Intent intent = getIntent();
        String type = intent.getStringExtra("TYPE");

        TextView title = findViewById(R.id.tvTest);
        title.setText(type);

        switch (type) {
            case Burger: {
                String desc1 = "Our Signature Whooper with 7 layers between the buns in convenient size.";
                String desc2 = "Our best-selling burger with crispy veg patty,fresh onion and our signature sauce.";
                String desc3 = "Enjoy rich makhani flavour with mix veg patty topped up with fresh onion.";
                String desc4 = "Our best-selling burger with crispy veg double patty,fresh onion and our signature sauce.";
                String desc5 = "Tikki bhi,Twist bhi!! Our new signature burger with spicy sauce.";

                list.add(new OrderItems(burger1, desc1, burger1price, burger1id));
                list.add(new OrderItems(burger2, desc2, burger2price, burger2id));
                list.add(new OrderItems(burger3, desc3, burger3price, burger3id));
                list.add(new OrderItems(burger4, desc4, burger4price, burger4id));
                list.add(new OrderItems(burger5, desc5, burger5price, burger5id));
                break;
            }

            case Momos: {
                String title1 = "Cheese Steamed Momos";
                String title2 = "Mix Veg Steamed Momos";
                String title3 = "Mix Veg Fried Momos";
                String title4 = "Cheese Fried Momos";
                String title5 = "Tandoori Gravy Momos";
                String desc1 = "Fine dough stuffed with cheese packed with our special sauce.";
                String desc2 = "Veggies stuffed in fine dough and steamed.";
                String desc3 = "Enjoy fried momos served with our signature sauce.";
                String desc4 = "Our best-selling momos with cheese and fried for best-in-class taste.";
                String desc5 = "Crunchy momos served with our signature sauce.";
                String price1 = "Rs.50";
                String price2 = "Rs.45";
                String price3 = "Rs.50";
                String price4 = "Rs.70";
                String price5 = "Rs.60";
                list.add(new OrderItems(title1, desc1, price1, R.drawable.cheese_setamed));
                list.add(new OrderItems(title2, desc2, price2, R.drawable.mix_veg_momos));
                list.add(new OrderItems(title3, desc3, price3, R.drawable.mix_veg_fried));
                list.add(new OrderItems(title4, desc4, price4, R.drawable.cheese_fried));
                list.add(new OrderItems(title5, desc5, price5, R.drawable.tandoori_gravy));
                break;
            }

            case PaneerTikka: {

            }
        }
        adapter = new OrdersAdapter(list,this);

        RecyclerView recyclerView = findViewById(R.id.recyclerOrders);
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));


    }

    @Override
    public void onOrderClickk(int position) {
        final BottomSheetDialog bottomSheetDialog = new BottomSheetDialog(this);
        bottomSheetDialog.setContentView(R.layout.orders_bottom_sheet);

        TextView name , price , total ;
        ImageView image;
        Button buy = bottomSheetDialog.findViewById(R.id.btnBuyBottomSheetOrders);
        name = bottomSheetDialog.findViewById(R.id.tvItemNameBottomSheetOrders);
        price = bottomSheetDialog.findViewById(R.id.tvItemPriceBottomSheetOrders);
        total = bottomSheetDialog.findViewById(R.id.tvTotalPriceBottomSheetOrders);
        image= bottomSheetDialog.findViewById(R.id.ivBottomSheetOrders);

        switch (position) {
            case 0 : {
                assert name != null;
                assert price != null;
                assert total != null;
                assert image != null;
                name.setText(burger1);
                price.setText(burger1price);
                total.setText(burger1price);
                image.setImageResource(burger1id);
                break;
            }
            case 1 : {
                assert name != null;
                assert price != null;
                assert total != null;
                assert image != null;
                name.setText(burger2);
                price.setText(burger2price);
                total.setText(burger2price);
                image.setImageResource(burger2id);
                break;
            }
            case 2 : {
                assert name != null;
                assert price != null;
                assert total != null;
                assert image != null;
                name.setText(burger3);
                price.setText(burger3price);
                total.setText(burger3price);
                image.setImageResource(burger3id);
                break;
            }
            case 3 : {
                assert name != null;
                assert price != null;
                assert total != null;
                assert image != null;
                name.setText(burger4);
                price.setText(burger4price);
                total.setText(burger4price);
                image.setImageResource(burger4id);
                break;
            }
            case 4 : {
                assert name != null;
                assert price != null;
                assert total != null;
                assert image != null;
                name.setText(burger5);
                price.setText(burger5price);
                total.setText(burger5price);
                image.setImageResource(burger5id);
                break;
            }
            default:
                Toast.makeText(this, "Error", Toast.LENGTH_SHORT).show();
        }

        assert buy != null;
        buy.setOnClickListener(view -> {
            Toast.makeText(this, "Order Placed Successfully !", Toast.LENGTH_SHORT).show();
        });
        bottomSheetDialog.show();
    }
}

File: app/src/main/java/the/mrsmile/foodorderer/activities/OrderActivity.kt
package the.mrsmile.foodorderer.activities

import android.graphics.Color
import androidx.appcompat.app.AppCompatActivity
import the.mrsmile.foodorderer.adapters.OrdersAdapter.OnOrderClick
import the.mrsmile.foodorderer.adapters.OrdersAdapter
import the.mrsmile.foodorderer.R
import android.os.Bundle
import android.view.View
import android.widget.*
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.ktx.auth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import com.google.firebase.database.ktx.database
import com.google.firebase.database.ktx.getValue
import com.google.firebase.ktx.Firebase
import the.mrsmile.foodorderer.database.Dao
import the.mrsmile.foodorderer.databinding.ActivityOrderBinding
import the.mrsmile.foodorderer.databinding.OrdersBottomSheetBinding
import the.mrsmile.foodorderer.models.BagItems
import the.mrsmile.foodorderer.models.OrderActivityItems
import java.util.ArrayList

class OrderActivity : AppCompatActivity(), OnOrderClick {

    private lateinit var binding: ActivityOrderBinding
    private var adapter: OrdersAdapter? = null
    private var type: Int? = null
    private lateinit var PATH: String
    private lateinit var dao: Dao
    private lateinit var daoBag: Dao
    private lateinit var auth: FirebaseAuth
    private var listOrder: ArrayList<OrderActivityItems> = ArrayList()
    private var listBagGlobal = ArrayList<BagItems>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityOrderBinding.inflate(layoutInflater)
        setContentView(binding.root)

        transparentStatusBar()
        auth = Firebase.auth
        val currentUser = auth.currentUser
        daoBag = Dao(
            Firebase.database.getReference(currentUser?.uid!!)
                .child(BagItems::class.java.simpleName)
        )

        type = intent.getIntExtra("TYPE", -1)
        if (type != -1) PATH = "${OrderActivityItems::class.java.simpleName}/$type"
        loadBigImage()
        getData()
        getBagItems()
    }

    override fun onOrderClickk(position: Int) {
        showOrderBottomSheet(position)
    }

    private fun getData() {
        dao = Dao(Firebase.database.getReference(PATH))
        val list: ArrayList<OrderActivityItems> = ArrayList()
        val valueEventListener = object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (snapshot.exists()) {
                    for (dataSnapshot in snapshot.children) {
                        val item = dataSnapshot.getValue(OrderActivityItems::class.java)
                        if (!list.contains(item) && item != null) {
                            list.add(item)
                        }
                    }
                    setRecyclerView(list)
                    listOrder = list
                }
            }

            override fun onCancelled(error: DatabaseError) {
                Snackbar.make(binding.root, "Something went wrong!", Snackbar.LENGTH_SHORT)
                    .show()
            }

        }
        dao.get().addValueEventListener(valueEventListener)
    }

    private fun setRecyclerView(list: ArrayList<OrderActivityItems>) {

        adapter = OrdersAdapter(list, this, this)
        val recyclerView = findViewById<RecyclerView>(R.id.recyclerOrders)
        recyclerView.adapter = adapter
        recyclerView.layoutManager = LinearLayoutManager(this)
        binding.progressBarOrdersActivity.hide()
    }

    private fun loadBigImage() {
        Firebase.database.getReference("OrderActivityImages").child(type.toString()).child("IMAGE")
            .addValueEventListener(object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {
                    if (snapshot.exists()) {
                        val item = snapshot.getValue<String>()
                        if (item != null)
                            Glide.with(this@OrderActivity).load(item).centerCrop()
                                .into(binding.ivOrdersActivity)
                    }
                }

                override fun onCancelled(error: DatabaseError) {
                    Snackbar.make(binding.root, "Something went wrong!", Snackbar.LENGTH_SHORT)
                        .show()
                }
            })
    }

    private fun showOrderBottomSheet(position: Int) {
        val bottomSheetDialog = BottomSheetDialog(this)
        val sheetBinding = OrdersBottomSheetBinding.inflate(layoutInflater)
        bottomSheetDialog.setContentView(sheetBinding.root)


        val title = sheetBinding.tvItemNameBottomSheetOrders
        val desc = sheetBinding.tvItemDescBottomSheetOrders
        val image = sheetBinding.ivBottomSheetOrders
        val price = sheetBinding.tvPriceBottomSheetOrders
        val btnPlus = sheetBinding.btnCounterPlusBottomSheet
        val btnMinus = sheetBinding.btnCounterMinusBottomSheet
        val buy = sheetBinding.btnBuyBottomSheetOrders
        val counter = sheetBinding.tvCounterTextBottomSheet
        var count = 1
        Glide.with(this).load(listOrder[position].image).placeholder(R.drawable.placeholder2)
            .into(image)
        title.text = listOrder[position].title
        desc.text = listOrder[position].desc
        price.text = listOrder[position].price.toString()

        btnPlus.setOnClickListener {
            count++
            counter.text = count.toString()
        }
        btnMinus.setOnClickListener {
            if (count > 1) {
                count--
                counter.text = count.toString()
            }
        }

        buy.setOnClickListener {
            val name = listOrder[position].title
            val pricee = listOrder[position].price
            val imagee = listOrder[position].image
            val time = listOrder[position].time
            val totalAmount = pricee?.times(count)
            val item = (BagItems(time, name, count, imagee, totalAmount))

            var flag = false
            var pos = 0
            if (listBagGlobal.isNotEmpty()) {
                for (i in 0 until listBagGlobal.size) {
                    if (listBagGlobal[i].name == item.name) {
                        flag = true
                        pos = i
                    }
                }
                if (flag) {
                    val quantityNew =
                        item.quantity?.let { it1 ->
                            listBagGlobal[pos].quantity?.plus(it1)
                        }
                    val priceNew = item.price?.let { it2 ->
                        listBagGlobal[pos].price?.plus(it2)
                    }
                    item.quantity = quantityNew
                    val hashMap = HashMap<String, Any>()
                    val key = listBagGlobal[pos].key
                    if (quantityNew != null && priceNew != null) {
                        hashMap["quantity"] = quantityNew
                        hashMap["price"] = priceNew
                    }
                    if (key != null) {
                        daoBag.update(key, hashMap).addOnCompleteListener {
                            if (it.isSuccessful) {
                                showToast()
                            }
                        }
                    }
                } else {
                    daoBag.addBagItem(item).addOnCompleteListener {
                        if (it.isSuccessful) {
                            showToast()
                        }
                    }
                }

            } else if (listBagGlobal.isNullOrEmpty()) {
                daoBag.addBagItem(item).addOnCompleteListener {
                    if (it.isSuccessful) {
                        showToast()
                    }
                }
            }
            bottomSheetDialog.dismissWithAnimation = true
            bottomSheetDialog.dismiss()
        }
        bottomSheetDialog.show()
    }

    private fun transparentStatusBar() {
        window?.decorView?.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN)
        window.statusBarColor = Color.TRANSPARENT
    }

    private fun getBagItems() {
        val list = ArrayList<BagItems>()
        val valueEventListener = object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (snapshot.exists()) {
                    for (dataSnapshot in snapshot.children) {
                        val item = dataSnapshot.getValue(BagItems::class.java)
                        if (dataSnapshot.key != null) {
                            item?.key = dataSnapshot.key
                        }
                        if (item != null) {
                            list.add(item)
                        }
                    }
                    listBagGlobal = list
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        }
        daoBag.get().addValueEventListener(valueEventListener)
    }

    private fun showToast() {
        val view =
            layoutInflater.inflate(R.layout.custom_toast, findViewById(R.id.llCustomToast), false)
        val toast = Toast(this)
        toast.duration = Toast.LENGTH_SHORT
        toast.view = view
        toast.show()
    }
}